src/node_report.cc

```diff
@@ -0,0 +1,1024 @@
+  MallocedBuffer<char> buffer(buffer_size);
+  int rc = -1;
+  while (rc != 0 && buffer_size < 1024 * 1024) {
+    rc = loadquery(L_GETINFO, buffer.data, buffer_size);
+    if (rc == 0) {
+      break;
+    }
+    buffer_size += buffer_inc;
+    buffer = MallocedBuffer<char>(buffer_size);
+  }
+  if (rc == 0) {
+    char* buf = buffer.data;
+    ld_info* cur_info = nullptr;
+    json_arraystart(out, *indent, "shared_objects");
+    json_indent(indent);
+    do {
+      std::ostringstream str;
+      cur_info = reinterpret_cast<ld_info*>(buf);
+      char* member_name = cur_info->ldinfo_filename +
+          strlen(cur_info->ldinfo_filename) + 1;
+      if (*member_name != '\0') {
+        str << cur_info->ldinfo_filename << "(" << member_name << ")";
+        json_arrayentrynormal(out, *indent, str.str());
+        str.str("");
+      } else {
+        json_arrayentrynormal(out, *indent, cur_info->ldinfo_filename);
+      }
+      buf += cur_info->ldinfo_next;
+    } while (cur_info->ldinfo_next != 0);
+    unread(out);
+    json_deindent(indent);
+    json_arrayendfinal(out, *indent);
+  }
+#elif __sun
+  Link_map* p;
+
+  json_arraystart(out, *indent, "shared_objects");
+  json_indent(indent);
+  if (dlinfo(RTLD_SELF, RTLD_DI_LINKMAP, &p) != -1) {
+    for (Link_map* l = p; l != nullptr; l = l->l_next) {
+      // out << "  " << l->l_name << "\n";
+      json_arrayentrynormal(out, *indent, l->l_name);
+    }
+    unread(out);
+    json_deindent(indent);
+    json_arrayendfinal(out, *indent);
+  }
+
+#elif _WIN32
+  // Windows implementation - get a handle to the process.
+  HANDLE process_handle = OpenProcess(PROCESS_QUERY_INFORMATION|PROCESS_VM_READ,
+                                      FALSE, GetCurrentProcessId());
+  if (process_handle == nullptr) {
+    return;
+  }
+  // Get a list of all the modules in this process
+  DWORD size_1 = 0, size_2 = 0;
+  // First call to get the size of module array needed
+  if (EnumProcessModules(process_handle, nullptr, 0, &size_1)) {
+    MallocedBuffer<HMODULE> modules(size_1);
+
+    // Second call to populate the module array
+    if (EnumProcessModules(process_handle, modules.data, size_1, &size_2)) {
+      json_arraystart(out, *indent, "shared_objects");
+      json_indent(indent);
+      for (int i = 0;
+           i < (size_1 / sizeof(HMODULE)) && i < (size_2 / sizeof(HMODULE));
+           i++) {
+        TCHAR module_name[MAX_PATH];
+        // Obtain and print the full pathname for each module
+        if (GetModuleFileNameEx(process_handle,
+                                modules.data[i],
+                                module_name,
+                                sizeof(module_name) / sizeof(TCHAR))) {
+          json_arrayentrynormal(out, *indent, module_name);
+        }
+      }
+      unread(out);
+      json_deindent(indent);
+      json_arrayendfinal(out, *indent);
+    }
+  }
+
+  // Release the handle to the process.
+  CloseHandle(process_handle);
+#endif
```

Comment:
It might be nice to split this whole function into a separate API à la `NativeSymbolDebuggingContext` that isn’t concerned with the JSON formatting…?

Reply:
sorry, did not follow this - the section is basically to enumerate over the loaded libraries, not for debugging context. Can you please elaborate? I may be missing something.